scene/resources/2d/tile_set.cpp

```diff
@@ -6486,21 +6486,38 @@ Ref<NavigationPolygon> TileData::get_navigation_polygon(int p_layer_id, bool p_f
 		Ref<NavigationPolygon> transformed_polygon;
 		transformed_polygon.instantiate();
 
-		PackedVector2Array new_points = get_transformed_vertices(layer_tile_data.navigation_polygon->get_vertices(), p_flip_h, p_flip_v, p_transpose);
```

Comment:
This line was good, the same vertex/point belonging to many different polygons transforms to the same point for each such transformed polygon. So no duplicated vertices need to be made (like in the new code below).

The issue was how the index array is created, as it assumed that all vertices form a single polygon. So a proper fix would be to just copy the indexes for all relevant polygons.

Hence I suggest:
- reverting the removal of this line,
- removing the added code below, and just copy the existing polygons' indices.

Looking at the NavigationPolygon's API **conceptually (*)** it should be as simple as doing `transformed_polygon.set_polygons(layer_tile_data.navigation_polygon->get_polygons())`. That's in case the winding order of the polygons' vertices doesn't matter, which seems to be the case because the currently proposed code also doesn't preserve the winding order and is said to work just fine.

If the winding order was meant to be preserved then something like this would need to be done (the winding order changes due to the transpose/flip_v/flip_h transformation if either 1 or 3 of these flags are set):
```cpp
bool need_to_reverse_indices = (p_flip_h != p_flip_v != p_transpose);
if (need_to_reverse_indices) {
// Fill the index array in reversed order (assuming the order of individual polygons doesn't matter).
...
}
```

**(*)** The interesting part is... `NavigationPolygon::{set|get}_polygons` seems to be not defined/implemented, they're just declared. :thinking: :smile:
There's `NavigationPolygon::{_set|_get}_polygons` but they're `protected` and hence can't be used in here.
cc @smix8 for explaining what the intention was there / what should be done about these. :slightly_smiling_face:
https://github.com/godotengine/godot/blob/5241d30bfa223ed45784e32d8143d20a98a8d862/scene/resources/2d/navigation_polygon.h#L117-L118
https://github.com/godotengine/godot/blob/5241d30bfa223ed45784e32d8143d20a98a8d862/scene/resources/2d/navigation_polygon.h#L65-L66
https://github.com/godotengine/godot/blob/5241d30bfa223ed45784e32d8143d20a98a8d862/scene/resources/2d/navigation_polygon.cpp#L95-L114

So I'd say the first step is making something like `NavigationPolygon::{set|get}_polygons` available.

Reply:
I agree with your suggested fix. We should make `NavigationPolygon::{set|get}_polygons` available. Happy to propose a implementation if you agree/depending on the response of @smix8.

Reversing polygon order should - as far as my limited knowledge of the codebase goes - not be necessary. At least the triangulation algorithm reverses index order if necessary (https://github.com/godotengine/godot/blob/5241d30bfa223ed45784e32d8143d20a98a8d862/core/math/triangulate.cpp#L131). I think that we should probably still do it to avoid potential future issues.

Reply:
Sure you're free to propose an implementation - that's your PR in the end. :upside_down_face: So yes, you can push a new version if you like. Then you'd get the feedback directly for it etc., so it can even end up faster to resolve/merge in the end.

Regarding the winding order seems it doesn't matter: https://github.com/godotengine/godot/issues/85122#issuecomment-2100906971. So don't bother doing any reversing etc.

Reply:
The winding of the actual NavigationPolygon/Mesh vertices and polygon indices does not matter. The NavigationServer converts to a native polygon format anyway.

The winding would only matter for the NavigationPolygon `outlines`. If the NavigationPolygon has stacked inner outlines the polygon clipping operations can turn out wrong for hole calculations. There are many reasons why this would break in a messed up placement situation and winding is only the last straw so dont really bother.

That the `set_polygons()` and `get_polygons()` functions have no body is an oversight. They are a leftover from the 2D navigation mesh baking PR that added them. The function body got removed in a rebase / refactoring shakeup.

If they are added they should be added for both NavigationPolygon and NavigationMesh at the same time. The internal code that currently uses add_polygon() (e.g. NavMeshGenerator2D/3D) in a loop should then be improved to use the newer function where it makes sense.

Reply:
So @libklein I think the simplest would be to use `{add|get}_polygon`/`get_polygon_count` in this PR (like as mentioned by **smix8** other internal code currently do).

And then `{set|get}_polygons` part could be addressed in a separate PR (either by you, or whoever else want to do that).

Reply:
> If they are added as internal function fast-paths using Vector they should be added for both NavigationPolygon and NavigationMesh at the same time. The internal code that currently uses add_polygon() (e.g. NavMeshGenerator2D/3D) in a loop should then be improved to use the newer function where it makes sense. The functions dont need bindings, the binding system can not deal with nested Vector arrays and we already have the equivalents that use slow and garbled TypeArrays.

Concerning the types. I've looked at the definition of [`Polygon2D`](https://github.com/libklein/godot/blob/d95794ec8a7c362b06a9cf080e2554ef77adb667/scene/2d/polygon_2d.h#L39). It uses the `Array` to store Polygons. Is my assumption that this is to enable bindings to the function correct? 

I'd suggest removing the Polygon struct from https://github.com/godotengine/godot/blob/5241d30bfa223ed45784e32d8143d20a98a8d862/scene/resources/2d/navigation_polygon.h#L41. It is not exposed in the public facing API and just forces unnecessary type casting/conversion in set/get methods.